### DISCUSSION TITLE
What Priority means in Jakarta?

### DISCUSSION THREAD

**Discussion by Emily-Jiang** [eclipse-ee4j/config#8] [General]
Further to last week's discussion, I spent a bit more time looking at the usage in CDI and try to recall what we discussed in the past regarding priority understanding. For priority used for overwrite reasons, such as converter and configsources.

We mainly discussed the invocation order for interceptors etc. No example was found on overwrite purpose as we used in Config.

I did further digging and found in CDI spec, the following explains clearly that highest priority wins.

https://jakarta.ee/specifications/cdi/3.0/jakarta-cdi-spec-3.0.html#unsatisfied_and_ambig_dependencies

All the beans left are alternatives with a priority, or producer methods or fields of beans that are alternatives with a priority, then the container will determine the highest priority value, and eliminate all beans, except for alternatives with the highest priority and producer methods and fields of alternatives with the highest priority value. If there is exactly one bean remaining, the container will select this bean, and the ambiguous dependency is called resolvable.

The above sentence clearly says the highest priority wins. MP Config priority is consistent with this behaviour.

**Comment by joakime** [comment]
The `@Priority` feature of CDI is great and all.
However, Jakarta Config, does not need to do anything to allow CDI to use it.
The priority concepts and layers we have discussed (in the past) in Jakarta Config are just not necessary to exist in Jakarta Config.
CDI + Priority will still work.

**Reply by Emily-Jiang** [reply]
> The priority concepts and layers we have discussed (in the past) in Jakarta Config are just not necessary to exist in Jakarta Config.
 
The priority concept needs to exist in Jakarta Config to specify the order of config source and converters.

**Reply by joakime** [reply]
I disagree, this priority is just a crutch your specific implementation of Config has implemented.

Others (like Eclipse Jetty, Google, etc) will base the order on other things, such as a topological sort from enabled feature sets.

It should not be present in Config itself.

**Comment by jansupol** [comment]
In Jakarta REST, the [Spec](https://jakarta.ee/specifications/restful-ws/3.0/jakarta-restful-ws-spec-3.0.html#priorities) says: 
> Execution chains for extension points `ContainerRequest`, `PreMatchContainerRequest`, `ClientRequest`, `ReadFrom` and `WriteTo`
are sorted in ascending order; the lower the number the higher the priority. Execution chains for extension
points `ContainerResponse` and `ClientResponse` are sorted in descending order; the higher the number the
higher the priority. These rules ensure that response filters are executed in reversed order of request filters.

For the request  _the lower the number the higher the priority_. It has the following consequences:
1. If only one is to be used, the one with the lowest number wins
2. If more are used, the one with the highest number is used the first, the one with the lowest number is used the last so that it can override all previous and win.

**Reply by Emily-Jiang** [reply]
> In Jakarta REST, the [Spec](https://jakarta.ee/specifications/restful-ws/3.0/jakarta-restful-ws-spec-3.0.html#priorities) says:
> 
> > Execution chains for extension points `ContainerRequest`, `PreMatchContainerRequest`, `ClientRequest`, `ReadFrom` and `WriteTo`
> > are sorted in ascending order; the lower the number the higher the priority. Execution chains for extension
> > points `ContainerResponse` and `ClientResponse` are sorted in descending order; the higher the number the
> > higher the priority. These rules ensure that response filters are executed in reversed order of request filters.
> 

In the above statement, it contradicts each other. One says the lower number, the higher priority. Later says the higher number the higher priority.

> For the request _the lower the number the higher the priority_. It has the following consequences:
> 
>     1. If only one is to be used, the one with the lowest number wins
> 
>     2. If more are used, the one with the highest number is used the first, the one with the lowest number is used the last so that it can override all previous and win.

Again, this contradicts with what interceptor specs says:


    Priorities that define the order in which interceptors are invoked. These values are intended to be used with the Priority annotation for interceptors that are defined by means of interceptor binding.

    Interceptors with smaller priority values are called first. If more than one interceptor has the same priority, the relative order of those interceptors is undefined.

When we discusses the priority in CDI group a while ago, we noticed JAX-RS has odd behaviour. However, I think CDI did it right: sorting interceptors based on the priority in ascending order. For Alternative, the higher priority the winner. 
Therefore, for priority, for queuing purpose, it should use ascending order. For override purpose, the higher priority, the final winner.

**Reply by jansupol** [reply]
> Again, this contradicts with what interceptor specs

It looks like it. Jakarta REST 4.0 (Jakarta EE 11) is planned to be very much CDI dependent, and IMO the CDI version of  `@Priority` behavior would better be accepted. Hence, I would prefer it the CDI way in Jakarta Config.

**Reply by joakime** [reply]
Per the restful-ws TCK the `@jakarta.annotation.Priority` annotation is used to layer extensions points between the REST Client and the REST Endpoint.

`REST Client -> [extensions] -> REST endpoint`

The difference you see in what they consider important is due to how the request vs the response flows through the list of extensions

Highest value is closest to REST endpoint side, the most important.

<pre>
  ExtFirst registered at priority 10
  ExtSecond registered at priority 20
  ExtThird registered at priority 30

  Request  from REST   Client -> ExtFirst -> ExtSecond -> ExtThird -> REST Endpoint
  Response from REST Endpoint -> ExtThird -> ExtSecond -> ExtFirst -> REST Client
</pre>

restful-ws recommends that you use the predefined values from [`jakarta.ws.rs.Priorities`](https://github.com/eclipse-ee4j/jaxrs-api/blob/3.0.0/jaxrs-api/src/main/java/jakarta/ws/rs/Priorities.java) as the value for `@jakarta.annotation.Priority(int)`.

If you look at the predefined priority list, you'll understand this order better. (do authn/authz first, then headers, then decoders/encoders, then user).

Looking back the history of restful-ws, you'll see that this entire layer exists because they were unhappy with the lack of control they have with the servlet spec.
So they implemented a decent chunk of the servlet spec behaviors themselves just to exert this control they desired. (servlet auth layer + servlet filter order + late config/binding)
Once Jakarta Core HTTP exists, this entire layer in restful-ws has no reason exist anymore.

**Reply by jansupol** [reply]
I stand corrected, thank you @joakime. My memory betrayed me. To sum the behaviour (sorry for confusion):

1. If only one is used, the one with the lowest number wins
2. If more are used, the one with the lowest number is used the first, the one with the highest number is used the last.

**Reply by Emily-Jiang** [reply]
It still suggests the higher priority the most important. Whatever called last will have the final say. CDI is very clear that highest priority wins for alternative beans.

**Reply by Emily-Jiang** [reply]
I discussed the priority usage in CDI weekly call this week.
For CDI:
1. If only one is used, the one with the highest number wins (e.g. alternatives)
2. If more are used, the one with the lowest number is used the first, the one with the highest number is used the last (e.g. CDI interceptors)

**Reply by struberg** [reply]
As written on the list: JAX-RS sadly uses both directions. There are cases where the highest value in `@Priority` wins and cases where the lowest value wins. Don't ask me why, but that's the way it is. Also the JAX-RS spec did use Priority a long time after it's introduction. So probably we should not take care too much about JAX-RS at all?

**Comment by joakime** [comment]
Here's a breakdown of a specific setup with servlet, but without Application created `ConfigSource` instances (I can flesh this out for that as well)

```
JVM
  Distribution
    libs/
      lib-foo.jar!/META-INF/config.properties (lib specific, has dependencies)
      lib-bar.jar!/META-INF/config.properties (lib specific, has dependencies)
    modules/
      technology-foo/
        config.properties (technology specific, has dependencies)
      technology-bar/
        config.properties (technology specific, has dependencies)
    default/
      config-a.properties (dist specific, has dependencies)
      config-b.properties (dist specific, has dependencies)
  Instance
    libs/
      lib-zed.jar!/META-INF/config.properties (lib specific, has dependencies)
    config.d/
      technology.properties (technology dependencies)
      tech-a.properties (instance specific, has dependencies)
      tech-b.properties (instance specific, has dependencies)
    webapps/
      bar-b.war  
        WEB-INF/
          config.properties (webapp specific properties)
          lib/
            flax.jar!/META-INF/config.properties (lib specific)
      bar-b.xml
        <config> section
```

**Comment by ljnelson** [comment]
The root question behind all this priority and ordinal talk is: will Jakarta Config require a _total_ (not partial) order of configuration value sources within an application?  It seems like that's what [`getOrdinal()` on MicroProfile Config's `ConfigSource`](https://javadoc.io/static/org.eclipse.microprofile.config/microprofile-config-api/2.0/org/eclipse/microprofile/config/spi/ConfigSource.html#getOrdinal--) is trying to help with, but of course it cannot by itself ever guarantee a total order (total ordering of any application components within an application is definitionally an application-level concern, not a component-level concern; your `ConfigSource` and my `ConfigSource` can happily report the same ordinal of, say, 37 and are now considered indistinguishable; this is a partial order).  Maybe Jakarta Config only _wants_ a partial order, in which case: is there really a point to it, given that in such a case any configuration value source can be indistinguishable from any other?

**Reply by joakime** [reply]
on web profile and servlet, configuration is layered as well.
some comes from startup, some comes later during deployment.
a "total order" is too harsh of requirement.

**Reply by ljnelson** [reply]
I'm not very familiar with modern Servlet stuff.  If there are two sources of configuration data that are indistinguishable, how does Servlet pick one over the other?  (Or is that a nonsensical question, because, say, Servlet doesn't permit there to be indistinguishable configuration data sources (unlike, say, MicroProfile Config)?)  A total ordering of elements just means each element is distinguishable from every other element; that doesn't seem very harsh.  (I am also not arguing for or against it.)  I would think that even with Servlet if I start the same Servlet-based application up several times, I would get the same sources of configuration data being chosen each time, yes?

**Reply by joakime** [reply]
There's the Container, the Instance, and each WebApp.
Or said another way.  System/Server classloader, API classloader, and WebApp classloader.  There are clear and well defined rules for behavior for each per the Servlet spec.
Specs like JSP and WebSocket tend to lean on those spec behaviors as well, but are not constrained by them, often adding more rules to the individual layers.
Each one should be considered a layer of ConfigSources, with each layer above it being the layer below + additional configuration.

**Reply by ljnelson** [reply]
And within each layer I assume it's completely deterministic, yes?  That is, within a layer it's impossible (or nonsensical) to have two sources of configuration data that are indistinguishable?

**Reply by joakime** [reply]
Its possible to have conflicts.  2 different technologies/specs/apis could have similar needs but different configurations.  solving for that is 2 stage.  once for the webapp artifact / deployable itself (the build-time fix), which is often enough, but sometimes also providing configuration to solve the conflict outside of the webapp can creep up as well (the production / env fix).

**Reply by ljnelson** [reply]
I think we're talking past each other.  Let me approach it a different way: right now, in a MicroProfile Config-2.0-using application, it is quite possible to put two `ConfigSource`s on the classpath with the same `ordinal` and `name` values, with wildly different `getValue()` semantics, and with no implementation-independent way to override those properties.  Consequently, the application's behavior is unpredictable.  Is the same sort of thing true of Servlet's configuration mechanism?  I would wager no: there is probably a way to resolve conflicts at deployment time that doesn't involve recompiling code.  Is that true?

**Reply by joakime** [reply]
Any issue like you describe are Spec/API specific.

For conflicts specific to Servlet Spec/API, this would either be handled by the packager of the WAR at development time.
Or (more commonly) it would be Servlet Implementation (Jetty, Tomcat, Glassfish, etc) specific fix (to override behaviors in the webapp from configuration outside of the webapp).

Some of the places people can change things.

1. Server / System level (this is the container itself) - changes here apply to the core functionality of the container (eg: how http works, features available, etc), and (generally speaking) has an impact on all user code running on the container.
2. API level - this is the Spec / API at a specific version - changes here apply to all webapps that use the same versions of the API - this is often done as a configuration on the instance of the container.
3. WebApp level - this is the specific webapp, at the specific context - changes here are bundled in the WAR deployable artifact and are changes made at development time
4. WebApp override - this is Web Container Implementation Specific - changes here apply to a specific WebApp deployable to a specific context - this is where changes by operations are usually made (the reasons for this can be numerous, from simple environment specific details, context specific details, production changes, backup regimes, failover policies, A/B testing behaviors, etc..) - configuration at this point is forced on the webapp, so that internal behaviors of the webapp don't even know it's happening, lots of different techniques are used here.

As for what this configuration looks like, right now it's a combination of descriptors, fragments, behaviors during webapp init, JNDI, internal sources of configuration, etc (long list).

It is the Web Container's responsibility to assemble the various layers and go through the initialization of each technology to (attempt to) create a reliable and reproducible resulting webapp on a specific context.
The various technologies in use have different rules and behaviors dictated by their Spec: Servlet, vs JSP, vs WebSocket have different rules at the different levels indicated above, when there a Spec conflict in these rules or something vague it's typically identified and handled between the specs for a resolution in a future revision.

A library using Spec Foo could be in a Web-Fragment that uses Spec Bar which is bundled in a WebApp.
If each of those libraries use ordinal, they have failed, as the order is not determined by them, only the Web Container can determine the order, and typically that is done based on the rules of initialization of the webapp.

**Reply by ljnelson** [reply]
Good, so there is a total order of configuration data sources.  Thanks.

**Reply by joakime** [reply]
@ljnelson the order exists, is determined only by the combination of implementation and spec rules, not the library code, not user code, not the ConfigSource itself, it exists transiently, is not fixed in stone, and does not exists with only one truth.

A developer could simply upgrade a single library version and now the order is different.
An overlay deployment (where 2 or more webapps are composited together) almost always changes the order.

Going from one technology to another introduces another layer which can influence the order.
HTTP upgrade to WebSocket.  (ConfigSources established based based on url path + uri template + and ws-sub-protocol)
Servlet (sees a ConfigSources from servlet init) -> include dispatch -> JSP (first execution sees potentially a new ConfigSources) -> jsp include -> JSF (first execution sees potentially new ConfigSources).

The "total order" is only true up to the finalization of a layer, but those layers are not known at startup of JVM or container, but come into existence based on other actions/events.

**Reply by ljnelson** [reply]
I completely understand. This is still a total order, not a partial one, i.e. there are deterministic rules that govern how otherwise indistinguishable configuration data sources are rendered distinguishable.

**Reply by joakime** [reply]
Not sure you can even say that there's a resolution of conflict in these technologies with regards to configuration.

* In the servlet spec, conflicts in configuration results in an "unavailable webapp" that requires admin intervention. (no traffic goes to it until the conflict is resolved and the webapp redeployed).  
   * Most groups choose to have the developers fix this, recompile, repackage, and redeploy.
   * A common alternative is Servlet Container Implementations (Jetty, Tomcat, etc) that have implementation specific techniques that can be used to resolve them without a recompilation / repackaging / redeploy.
* In websocket, a conflicting configuration results in a failed initialization and not serving any registered websocket, until the conflict is resolved and the websocket serving app is redeployed.  This almost always requires a recompile.
* In JSP, a configuration conflict results in the same, a failed state for all JSPs that rely on the failed configuration or JSP, other JSPs that don't depend on the failed one can still carry on.  (Most implementations have configurations for this behavior so that any failure results in total failure, vs limping state, most professional organizations only use precompiled JSP to identify these issues at development time).  To resolve the conflict requires a manual resolution, depending on how you use JSP it could be a simple in-place edit and dynamic recompile all the way up to a fix at dev side and redeploy.

**Reply by ljnelson** [reply]
Right. That means in Jakarta EE currently (or at least in the constellation of sub-specifications you reference above) it's impossible to start up and have random sources of configuration data chosen, because each sub-specification you reference above fails in some way when a total ordering of configuration data sources cannot be computed.  

By contrast, in MicroProfile Config (and now in Jakarta Config since it looks like the initial PR has been pushed), if there are two indistinguishable `ConfigSource`s in effect (mine returns `37` from `getOrdinal()` and "`config`" from `getName()` and yours does too, let's say, but each returns a different value from `getValue("something")`), it is undefined which of these `ConfigSource`s will be used.  Moreover, there is no currently specified way to correct this situation without recompiling code (or even to detect it in advance).

**Reply by joakime** [reply]
I'm still stuck on "total ordering".

If you have say ...

Layer 1: `ConfigSources` (id=1)
```
config-foo : (depends on bar)
config-bar : (depends on zed)
config-zed : (no dep)
```

and then a layer above introduces to the stack something that needs a new order.

Layer 2: `ConfigSources` (id=2)
```
config-foo : (depends on bar)
config-bar : (depends on zed)
config-lib : (depends on zed)
config-zed : (no dep)
```

This is a common scenario and in the overwhelmingly majority of use cases it does not introduce a conflict.
But the order is now different, and no ordinal was used to determine this.

If we had ordinal, the numbers change again.
You have two `ConfigSources` with different order and ordinals.
Only the `ConfigSources` (note plural) knows it, and not the `ConfigSource` (singular).
To the `ConfigSource` it has 2 different ordinals depending on the view into the configuration tree.

**Reply by ljnelson** [reply]
Perhaps we are in agreement.  I am using "total order" in the set-theory sense of the term: every element is distinguishable from every other element.

You've presented several scenarios that are governed by total ordering of configuration data sources.  That is: for any given instance of an application starting up, in your examples there is _an_ order that can be computed for that application's configuration data sources (I'm using abstract terms on purpose) that ensures that no randomness is present, i.e. for that application instance you can consult a rule book somewhere and determine which configuration data sources will be stacked in which order.  You've said as much above.

More importantly, if you start that very same application instance 100 times, there is apparently no chance that it will start with configuration data sources chosen at random.  Therefore, there is a total ordering of configuration data sources within that application instance.

Now, if you change the application's configuration and redeploy the application, you're running what amounts to a "new" application instance.  Likewise, for _this_ application instance, while the actual configuration data sources used may be different, there will still be a total ordering of them (every source will be distinguishable from every other source and there is a rule book that tells you definitively which ones will be used).  If you start _it_ 100 times, there is, likewise, no chance that it will start with configuration data sources organized in a different way each time.  This too is a total order of configuration data sources within _this_ application instance, just a different one.

By contrast, with MIcroProfile Config, and now with Jakarta Config, if your library ships with a `ConfigSource` that returns `37` as a hard-coded value from its `getOrdinal()` method and "`config`" as a hard-coded value from its `getName()` method, and my library ships with one that also returns `37` as a hard-coded value from its `getOrdinal()` method and "`config`" as a hard-coded value from its `getName()` method, but our respective `ConfigSource`s do wildly different things and return wildly different values, and our libraries containing them are combined into an application, the MicroProfile Config specification indicates our `ConfigSource`s must be sorted first by ordinal, then by name.  That's fine, but as you can see in this example even after this sorting, our `ConfigSource`s remain indistinguishable (according to the specification).  Hence MicroProfile Config requires only a _partial_ order of configuration data sources within an application, and if this hypothetical application containing these two sources is started 100 times, it is legal for the MicroProfile Config implementation being used to just…pick one at random each time.  We start the application two times, and the first time we get your `ConfigSource` trumping mine, and the second time mine trumps yours.  

This seems bad to me and worth changing in the Jakarta Config specification.  What is "bad" here is the partial ordering _within any given application instance_; it seems to me that Jakarta Config should have verbiage in it somewhere saying something like if two `ConfigSource`s within an application instance can't be distinguished for any reason during startup then an error should be raised.  Then at least the person putting the application together could arrange for a different `ConfigSource` with a higher ordinal to be used and to do something else…unless our sources' ordinals were already `Integer.MAX_VALUE` :-).  (With the current specification, there is no guaranteed way to be able to correct the problem.)

All of this is because the larger problem is that we're talking about ordering of configuration data sources (I'm using abstract terms on purpose) _within an application_, but the application's configuration ordering semantics are currently dictated _by the data sources_.  That leads inexorably to conflicts unless you can _also_ specify a means of resolving those conflicts when putting your application together.  Ideally you'd also be able to detect this state of affairs before trying things out.

**Reply by joakime** [reply]
> More importantly, if you start that very same application instance 100 times, there is apparently no chance that it will start with configuration data sources chosen at random. Therefore, there is a total ordering of configuration data sources within that application instance.

There is no guarantee of order for a ClassLoader.
Requesting content from the ClassLoader can result in different order with each execution.
This even impacts ServiceLoader.

This is also not a guarantee in the Servlet spec (when Servlet 3.0 introduced Servlet annotations and Web Fragments as concepts).
If you are a lazy developer and include `WEB-INF/lib/*.jar` content that has overlapping classes you are in an unfavorable place with repeatable/stable execution runs.
_(This is one of the reasons so many "duplicate class check" plugins exist in maven and gradle.)_
This also causes web-fragments to order differently during initialization as a result of the non-order of a ClassLoader.
There's some provisions in the web descriptor to order specific things before others, but it's highly limited and cannot compensate for many things.

Keep in mind that audience for Jakarta Config is the entire Jakarta EE ecosystem, not just MicroProfile.
The MicroProfile use cases are useful.
But those use cases don't have a special place or trump other use cases.

> What is "bad" here is the partial ordering within any given application instance; it seems to me that Jakarta Config should have verbiage in it somewhere saying something like if two ConfigSources within an application instance can't be distinguished for any reason during startup then an error should be raised. Then at least the person putting the application together could arrange for a different ConfigSource with a higher ordinal to be used and to do something else…unless our sources' ordinals were already Integer.MAX_VALUE :-). (With the current specification, there is no guaranteed way to be able to correct the problem.)

If Jakarta Config adopts ...

* `ConfigSources` a new class and is the responsibility of the Jakarta Config Implementation to create.
* `ConfigSource` without an ordinal.

Then we have a Jakarta Config that is adaptable for the MicroProfile case and also the other cases that can't / won't use ordinal.

**Reply by ljnelson** [reply]
OK; I'm now confused so I'll stop here. I think? maybe? I agree with you? that ideally within the Jakarta EE world there should always be a way to configure any application deterministically.  I hope I'm not saying anything other than that.  MicroProfile Config, and as of this writing Jakarta Config in its current state do not allow for this.  I am not a fan at all of either effort so far.  I'll leave it at that.

**Reply by struberg** [reply]
> There is no guarantee of order for a ClassLoader. Requesting content from the ClassLoader can result in different order with each execution. This even impacts ServiceLoader.

Yes, which is the reason we try to not rely on it by picking up ALL ConfigSources first, and then sort them via ordinal and names. If there are 2 ConfigSources with exactly the same name and ordinal, well then it will start to behave randomly. 
https://github.com/eclipse/ConfigJSR/blob/f15f4ee1ef9e68ce14af9b714899a94663afede3/api/src/main/java/javax/config/spi/ConfigSource.java#L112

If two ConfigSources have the same ordinal and name, AND contain the same config keys with different values, then non-portable behaviour results. We might also issue a warning or error in this case.

Also note that in practice this does not happen as frequent as one might think as implementations usually use the full classpath location (resource path or file path). We probably should explicitly mandate this for the default provided property file ConfigSources though!

**Reply by struberg** [reply]
> If Jakarta Config adopts ...
> 
>     * `ConfigSources` a new class and is the responsibility of the Jakarta Config Implementation to create.
> 
>     * `ConfigSource` without an ordinal.
> 
> 
> Then we have a Jakarta Config that is adaptable for the MicroProfile case and also the other cases that can't / won't use ordinal.

Could you probably elaborate on this idea a bit more in detail? I don't really get what the 2 should do and where they should differ. txs!

**Comment by dmlloyd** [comment]
We could probably unpin this now?

**Comment by ljnelson** [comment]
Unpinning.
